### PR TITLE
[Fix] Cast state-pool indices to int64 in fused_recurrent GDN update kernel

### PR DESCRIPTION
**Pull request by ChiachiHsieh**
## Motivation

`fused_recurrent_gated_delta_rule_update_fwd_kernel` derives three state-pool
offsets from caller-provided index tensors: the initial-state (`h0`) read, the
intermediate-state store, and the final-state write-back to `h0`. Each offset
multiplies the index by `HV * K * V` (or `cache_steps * HV * K * V`).

The indices are loaded in their caller dtype (int32 in all current callers) and
were never widened, so once a pool slot's row offset exceeds `2**31` the offset
silently wraps negative and the kernel reads/writes the wrong state row (or
triggers an illegal memory access). With unsharded GDN shapes the
intermediate-store stride is millions of elements per row, so this is reachable
within realistic pool sizes; tensor-parallel-sharded deployments stay below the
threshold, which is why it has been latent. The kernel's correctness should not
depend on the caller's shard geometry.

The same file already widens the `ssm_state_indices` loads to int64 (the packed
decode/update paths); these three loads were simply missed.

## Modifications

Widen the three index loads in `fused_recurrent_gated_delta_rule_update_fwd_kernel`
to int64 at the load (`.to(tl.int64)`), matching the existing `ssm_state_indices`
convention already used elsewhere in the file. There is no behavioral change for
any index that fits int32.

## Accuracy Tests

Added `test_large_state_pool_id_no_int32_overflow` in
`test/registered/jit/test_fused_verify_triton_gdn.py`: it runs identical physics
at slot id 0 and at a slot id whose flat offset exceeds `2**31`, and asserts
identical output and identical written-back state. Without the fix this triggers
a CUDA illegal memory access; with it, the test passes. The existing GDN
precision/decode tests continue to pass.

## Benchmarking and Profiling

Not a performance change — two integer-widening ops per program, off the hot
path.

## Checklist

- [x] Format your code according to the Code Formatting with Pre-Commit.
- [x] Add unit tests as outlined in the Running Unit Tests section.
- [ ] Update documentation as needed (N/A — kernel-internal fix).

---
### CI States

Latest PR Test (Base): :x: [Run #30340179060](https://github.com/sgl-project/sglang/actions/runs/30340179060)
Latest PR Test (Extra): :x: [Run #30340177887](https://github.com/sgl-project/sglang/actions/runs/30340177887)

### CODE REVIEW

File: python/sglang/kernels/ops/attention/fla/fused_recurrent.py [hunk 1]
```diff
@@ -897,7 +897,10 @@ def fused_recurrent_gated_delta_rule_update_fwd_kernel(
 
     b_h = tl.zeros([BV, BK], dtype=tl.float32)
     if USE_INITIAL_STATE:
-        idx = tl.load(h0_indices + i_n)
+        # Cast to int64 at the load: the row offset below multiplies this
+        # index by HV*K*V, which overflows int32 for large state-pool ids.
+        # (Same convention as the ssm_state_indices loads in this file.)
+        idx = tl.load(h0_indices + i_n).to(tl.int64)
```
[the remark this answers is not in the document]
Comment (ChiachiHsieh) [reply python/sglang/kernels/ops/attention/fla/fused_recurrent.py:903]: Good catch — done in a follow-up commit. Widened all three state-pool index loads (h0 read, intermediate-state store, and final write-back) in `fused_sigmoid_gating_recurrent.py` to int64, matching the `fused_recurrent.py` convention. The overflow test now also exercises this kernel through the fused sigmoid-gating path.

File: test/registered/jit/test_fused_verify_triton_gdn.py [hunk 2]
```diff
@@ -234,5 +234,65 @@ def test_mtp_single_step_decode(N: int):
...
+    )
+
+    # Reference at slot 0 (offset always fits int32).
+    ref_src = torch.randn(1, HV, K, V, dtype=torch.float, device="cuda")
+    ref_idx = torch.zeros(N, dtype=torch.int32, device="cuda")
+
+    # Same initial state at a high slot id whose flat offset overflows int32.
+    big_src = torch.zeros(big_id + 1, HV, K, V, dtype=torch.float, device="cuda")
+    big_src[big_id].copy_(ref_src[0])
+    big_idx = torch.full((N,), big_id, dtype=torch.int32, device="cuda")
```
[the remark this answers is not in the document]
Comment (ChiachiHsieh) [reply test/registered/jit/test_fused_verify_triton_gdn.py]: Done — the wrap test now allocates the high-id state buffer in `bfloat16` (~4.5 GB instead of ~9 GB), and frees it between the two kernel paths. The kernels accumulate in fp32 internally, so the storage dtype doesn't affect the overflow behaviour under test.
